tests/APIv2/AbstractResourceTest.php

```diff
@@ -140,7 +140,7 @@ public function testPatchFull() {
     public function testGetEdit($record = null) {
         if ($record === null) {
             $record = $this->record();
-            $row = $this->testPost();
+            $row = $this->testPost($record);
         } else {
             $row = $record;
         }
```

Comment:
I was unable to do this, it causes tests that override `testPost` to generate a record in a different way to fail. See https://travis-ci.com/vanilla/vanilla/jobs/169818468

Reply:
But I feel that problem is still somewhere else. And you are replacing one patch (to avoid the issue) with another one, but really fixing the issue.

What I see in report:

```
--- Expected
+++ Actual
@@ @@
 Array (
     'name' => null
-    'body' => 'Hello world!'
+    'body' => '&lt;p&gt;Hello world!&lt;/p&gt;\n
+    '
```
It shows that the value was changed from plain (or Markdown) text to html.
How the call of `$this->record()` inside `testGetEdit()` can fix that?

To me if there is a difference in tests behavior which depends on **when** we call the method `record()`, that is weird side effect. Probably I miss something. But this problem itself is pretty weird.

I will not block you and approve this PR. But I feel that we don't fix anything here :-(

Reply:
This is the way Vanilla works with body fields, specifically. On every response with the record, *except for a get-edit*, you'll get the formatted body (e.g. markdown converted to HTML). On a get-edit, you'll get the "raw" body. If you get the response of a post request and use that as the `$row`, it'll be a formatted body.  The value returned from the `record` method would be raw. That's why in `testGetEdit`, you need to get the raw contents from `record`, first, and then create a new post with them.

Reply:
@initvector I don't know what exactly you want to clarify. But yo just confuse me.

> you need to get the raw contents from record, first, and then create a new post with them.

That is what `$this->testPost();` does even without incoming param. It will **_get the raw contents from record,_** and **_and then create a new post with them._** 

 

```
public function testPost($record = null, array $extra = []) {
        $record = $record === null ? $this->record() : $record;
```

That is why I can not get how this new incoming parameter change the flow and create some extra "data pollution"

Reply:
@initvector I don't know what exactly you want to clarify. But you just confuse me.

> you need to get the raw contents from record, first, and then create a new post with them.

That is what `$this->testPost();` does even without incoming param. It will **_get the raw contents from record,_** and **_and then create a new post with them._** 

 

```
public function testPost($record = null, array $extra = []) {
        $record = $record === null ? $this->record() : $record;
```

That is why I can not get how this new incoming parameter change the flow and create some extra "data pollution"

Reply:
The `$record` in `testPost` is never exposed, so you can't see the "raw" input parameters used to create the record. What's returned from the method is what is returned from the post action on the resource. If you want to have the "raw" data, you'll need to pass it into `testPost`, instead of allowing that method to generate its own.

If you create a markdown comment, the "body" field of your input might look like "Hello **world**!". When you get back from the post request to the comments endpoint would have a rendered "body" field, so it might look like "Hello <strong>world</strong>!". If you want to get the "raw" comment body, you'd have to make a get request to /api/v2/comments/{id}/edit. The "body" in this response would be in the original markdown format (e.g. "Hello **world**!).

When you're trying to verify the result of a get-edit response, you often need to have the "raw" data you originally created the record with. Otherwise, you'll be trying to compare the "raw" and rendered body contents.

Reply:
@initvector Sorry, You are explaining 3rd or 4th time the same thing. And I still can't find the answer on my question. I feel like I understand every word you are saying. But I see 0 relation to my question and my point. 

Let me try to ask again.

We have this implemntation @charrondev just did:
```
 public function testGetEdit($record = null) {
        if ($record === null) {
            $record = $this->record();
            $row = $this->testPost($record);
        } else {
            $row = $record;
        }
        ...
```
 and  we have

```
public function testPost($record = null, array $extra = []) {
        $record = $record === null ? $this->record() : $record;
```

 I can not get how that is different from:
```
 public function testGetEdit($record = null) {
        if ($record === null) {
            // $record = $this->record();  <----- why we need to call this method outside? if we are in the same object instance
            $row = $this->testPost(); // <----- this method should instantiate  $record = $this->record(); absolutely the same way
        } else {
            $row = $record;
        }
        ...
```

We can continue offline when meet or whatever...

Reply:
This issue is almost exclusively related to our formatted fields, like "body". These fields go in as one format and come out as HTML in almost every request, except for a get-edit.

You need to have the raw post content, not the rendered content, to verify a get-edit response. If you're calling `testPost`, which returns the result of a post request, you're going to get the rendered HTML of a post. To avoid this, you call `record` and pass the result to `testPost`, so you can then verify the "body" field in the get-edit response matches your original request (from `record`).

A formatted field, like "body", will be HTML in a post request response. It will be the original raw format in a get-edit response. If you only use `testPost`, where the rendered body is returned, you'll never see the raw markdown (or whatever format). You cannot compare the body of a post request with the body of a get-edit response, because one is rendered and one isn't.

The methods in your code examples create the post the same, but the difference is you have access to the original raw (not rendered) body content in one and not in the other. You need the original raw body to validate the get-edit response. That's why there are assertions in this test on `$record` and not `$row`. We do not want to try to compare the rendered "body" field in a post response to the raw "body" field in a get-edit response.